File: apps/api/app/core/pipeline.py
from __future__ import annotations

from app.db.models import PipelineStatus

PIPELINE_GROUPS: dict[str, tuple[PipelineStatus, ...]] = {
    "ideation": (
        PipelineStatus.IDEA,
        PipelineStatus.RESEARCH,
        PipelineStatus.BRIEF,
    ),
    "scripting": (
        PipelineStatus.SCRIPT,
        PipelineStatus.REVIEW,
        PipelineStatus.APPROVED,
    ),
    "production": (
        PipelineStatus.READY_TO_SHOOT,
        PipelineStatus.SHOT,
        PipelineStatus.EDITING,
    ),
    "review": (
        PipelineStatus.REVIEW_EDIT,
        PipelineStatus.READY_TO_PUBLISH,
    ),
    "published": (
        PipelineStatus.PUBLISHED,
        PipelineStatus.ANALYTICS_REVIEW,
        PipelineStatus.REPURPOSE,
        PipelineStatus.ARCHIVED,
    ),
}

ALLOWED_TRANSITIONS: dict[PipelineStatus, set[PipelineStatus]] = {
    PipelineStatus.IDEA: {PipelineStatus.RESEARCH, PipelineStatus.ARCHIVED},
    PipelineStatus.RESEARCH: {PipelineStatus.BRIEF, PipelineStatus.IDEA, PipelineStatus.ARCHIVED},
    PipelineStatus.BRIEF: {PipelineStatus.SCRIPT, PipelineStatus.RESEARCH, PipelineStatus.ARCHIVED},
    PipelineStatus.SCRIPT: {PipelineStatus.REVIEW, PipelineStatus.BRIEF, PipelineStatus.ARCHIVED},
    PipelineStatus.REVIEW: {
        PipelineStatus.APPROVED,
        PipelineStatus.SCRIPT,
        PipelineStatus.ARCHIVED,
    },
    PipelineStatus.APPROVED: {PipelineStatus.READY_TO_SHOOT, PipelineStatus.SCRIPT},
    PipelineStatus.READY_TO_SHOOT: {PipelineStatus.SHOT, PipelineStatus.APPROVED},
    PipelineStatus.SHOT: {PipelineStatus.EDITING, PipelineStatus.READY_TO_SHOOT},
    PipelineStatus.EDITING: {PipelineStatus.REVIEW_EDIT, PipelineStatus.SHOT},
    PipelineStatus.REVIEW_EDIT: {PipelineStatus.READY_TO_PUBLISH, PipelineStatus.EDITING},
    PipelineStatus.READY_TO_PUBLISH: {PipelineStatus.PUBLISHED, PipelineStatus.REVIEW_EDIT},
    PipelineStatus.PUBLISHED: {PipelineStatus.ANALYTICS_REVIEW},
    PipelineStatus.ANALYTICS_REVIEW: {PipelineStatus.REPURPOSE, PipelineStatus.ARCHIVED},
    PipelineStatus.REPURPOSE: {PipelineStatus.IDEA, PipelineStatus.ARCHIVED},
    PipelineStatus.ARCHIVED: set(),
}


def compact_group(status: PipelineStatus) -> str:
    for name, statuses in PIPELINE_GROUPS.items():
        if status in statuses:
            return name
    return "ideation"

File: apps/web/src/lib/pipeline.ts
import type { PipelineGroup, PipelineStatus } from "@/lib/contracts";

export const PIPELINE_GROUPS: Record<PipelineGroup, PipelineStatus[]> = {
  ideation: ["idea", "research", "brief"],
  scripting: ["script", "review", "approved"],
  production: ["ready_to_shoot", "shot", "editing"],
  review: ["review_edit", "ready_to_publish"],
  published: ["published", "analytics_review", "repurpose", "archived"],
};

export const PIPELINE_LABELS: Record<PipelineStatus, string> = {
  idea: "Idea",
  research: "Research",
  brief: "Brief",
  script: "Script",
  review: "Script review",
  approved: "Approved",
  ready_to_shoot: "Ready to shoot",
  shot: "Shot",
  editing: "Editing",
  review_edit: "Edit review",
  ready_to_publish: "Ready to publish",
  published: "Published",
  analytics_review: "Analytics review",
  repurpose: "Repurpose",
  archived: "Archived",
};

export const NEXT_STATUS: Partial<Record<PipelineStatus, PipelineStatus>> = {
  idea: "research",
  research: "brief",
  brief: "script",
  script: "review",
  review: "approved",
  approved: "ready_to_shoot",
  ready_to_shoot: "shot",
  shot: "editing",
  editing: "review_edit",
  review_edit: "ready_to_publish",
  ready_to_publish: "published",
  published: "analytics_review",
  analytics_review: "repurpose",
  repurpose: "archived",
};

export function compactGroup(status: PipelineStatus): PipelineGroup {
  const match = (Object.entries(PIPELINE_GROUPS) as [PipelineGroup, PipelineStatus[]][]).find(
    ([, statuses]) => statuses.includes(status),
  );
  return match?.[0] ?? "ideation";
}
